pkg/workflows/ring/plugin.go

```diff
@@ -0,0 +1,290 @@
+	"github.com/smartcontractkit/libocr/commontypes"
+	"github.com/smartcontractkit/libocr/offchainreporting2plus/ocr3types"
+	"github.com/smartcontractkit/libocr/offchainreporting2plus/types"
+	"github.com/smartcontractkit/libocr/quorumhelper"
+)
+
+type Plugin struct {
+	mu sync.RWMutex
+
+	store  *Store
+	config ocr3types.ReportingPluginConfig
+	lggr   logger.Logger
+
+	batchSize     int
+	minShardCount uint32
+	maxShardCount uint32
+	timeToSync    time.Duration
+}
+
+var _ ocr3types.ReportingPlugin[[]byte] = (*Plugin)(nil)
+
+// ConsensusConfig holds the plugin configuration
+type ConsensusConfig struct {
+	MinShardCount uint32
+	MaxShardCount uint32
+	BatchSize     int
+	TimeToSync    time.Duration
+}
+
+// NewPlugin creates a consensus reporting plugin for shard orchestration
+func NewPlugin(store *Store, config ocr3types.ReportingPluginConfig, lggr logger.Logger, cfg *ConsensusConfig) (*Plugin, error) {
+	if cfg == nil {
+		cfg = &ConsensusConfig{
+			MinShardCount: 1,
+			MaxShardCount: 100,
+			BatchSize:     1000,
+			TimeToSync:    5 * time.Minute,
+		}
+	}
+
+	if cfg.MaxShardCount == 0 {
+		return nil, errors.New("max shard count cannot be 0")
+	}
+	if cfg.MinShardCount == 0 {
+		cfg.MinShardCount = 1
+	}
+	if cfg.BatchSize <= 0 {
+		cfg.BatchSize = 100
+	}
+	if cfg.TimeToSync <= 0 {
+		cfg.TimeToSync = 5 * time.Minute
+	}
+
+	return &Plugin{
+		store:         store,
+		config:        config,
+		lggr:          logger.Named(lggr, "RingPlugin"),
+		batchSize:     cfg.BatchSize,
+		minShardCount: cfg.MinShardCount,
+		maxShardCount: cfg.MaxShardCount,
+		timeToSync:    cfg.TimeToSync,
+	}, nil
+}
+
+func (p *Plugin) Query(_ context.Context, _ ocr3types.OutcomeContext) (types.Query, error) {
+	return nil, nil
+}
+
+func (p *Plugin) Observation(_ context.Context, _ ocr3types.OutcomeContext, _ types.Query) (types.Observation, error) {
+	shardHealth := p.store.GetShardHealth()
+
+	allWorkflowIDs := make([]string, 0)
+	for wfID := range p.store.GetAllRoutingState() {
+		allWorkflowIDs = append(allWorkflowIDs, wfID)
+	}
+	slices.Sort(allWorkflowIDs)
+
+	observation := &pb.Observation{
+		ShardHealthStatus: shardHealth,
+		WorkflowIDs:       allWorkflowIDs,
+		Now:               timestamppb.Now(),
+	}
+
+	return proto.MarshalOptions{Deterministic: true}.Marshal(observation)
+}
+
+func (p *Plugin) ValidateObservation(_ context.Context, _ ocr3types.OutcomeContext, _ types.Query, _ types.AttributedObservation) error {
+	return nil
+}
+
+func (p *Plugin) ObservationQuorum(_ context.Context, _ ocr3types.OutcomeContext, _ types.Query, aos []types.AttributedObservation) (quorumReached bool, err error) {
+	return quorumhelper.ObservationCountReachesObservationQuorum(quorumhelper.QuorumTwoFPlusOne, p.config.N, p.config.F, aos), nil
+}
+
+func (p *Plugin) Outcome(_ context.Context, outctx ocr3types.OutcomeContext, _ types.Query, aos []types.AttributedObservation) (ocr3types.Outcome, error) {
```

Comment:
In my PoC, the executions used the table, not workflows. I know we said it's fine for us to route based on workflows, but we still need to route individual execution ids and track where they went as I did in my example.

With that you have now, it's possible for a route request for workflow A with execution ID B to come in from node 0 just before we transition to `STEADY` state. It would be routed by workflow ID and land on the node from the prior steady state. If we then transition to `STEADY`, and node 1 get the request at that point (regardless of if it saw the state transition yet) we'll route it based on the new assignment for workflow A, which breaks thing.